### PR TITLE
[Parallel Iterators] Batching + Pipelining optimizations

### PR DESCRIPTION
**Pull request by amogkam**
## Why are these changes needed?
Adds optimizations to Parallel Iterators to improve throughput and.
* Adds internal batching for `gather_async`, `get_shard`, and `repartition` to reduce number of remote calls.
* Adds pipelining to `get_shard`.
* Fixes `StopIteration` error in `duplicate`
* Renames `async_queue_depth` in `gather_async` to `pipeline_queue_depth`

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/util/iter.py [hunk 1]
```diff
@@ -289,6 +291,9 @@ def repartition(self, num_partitions: int) -> "ParallelIterator[T]":
         Args:
             num_partitions (int): The number of shards to use for the new
                 ParallelIterator
+            batch_time (int): Batches items for batch_time milliseconds
+                on each shard before retrieving it.
+                Increasing batch_time reduces latency but improves throughput.
```
Comment (wuisawesome) [inline python/ray/util/iter.py:296]: I think you mean increases latency? 

Decreasing latency and improving throughput sounds great :)

**Comment by ericl** [comment]
Any benchmark results? Would be great to check in tests with reference results for future comparison.

File: python/ray/util/iter.py [hunk 2]
```diff
@@ -279,7 +280,8 @@ def local_shuffle(self, shuffle_buffer_size: int,
                 shuffle_buffer_size,
                 str(seed) if seed is not None else "None"))
 
-    def repartition(self, num_partitions: int) -> "ParallelIterator[T]":
+    def repartition(self, num_partitions: int,
+                    batch_time: int = 10) -> "ParallelIterator[T]":
```
Comment (ericl) [inline python/ray/util/iter.py:284]: Consider `batch_ms`

File: python/ray/util/iter.py [hunk 3]
```diff
@@ -415,16 +425,20 @@ def base_iterator(timeout=None):
         name = "{}.batch_across_shards()".format(self)
         return LocalIterator(base_iterator, SharedMetrics(), name=name)
 
-    def gather_async(self, async_queue_depth=1) -> "LocalIterator[T]":
+    def gather_async(self, batch_time: int = 10,
+                     pipeline_queue_depth: int = 1) -> "LocalIterator[T]":
```
Comment (ericl) [inline python/ray/util/iter.py:429]: Consider `num_async_requests` instead of pipeline_queue_depth.

File: python/ray/util/iter.py [hunk 4]
```diff
@@ -511,11 +530,23 @@ def shards(self) -> List["LocalIterator[T]"]:
         """Return the list of all shards."""
         return [self.get_shard(i) for i in range(self.num_shards())]
 
-    def get_shard(self, shard_index: int) -> "LocalIterator[T]":
+    def get_shard(self,
+                  shard_index: int,
+                  batch_time: int = 10,
+                  pipeline_queue_depth: int = 1) -> "LocalIterator[T]":
```
Comment (ericl) [inline python/ray/util/iter.py:536]: Consider `num_async_requests`.

**Review by ericl — CHANGES_REQUESTED** [review]

**Review by ericl — COMMENTED** [review]
I think we should go ahead and merge this, but with batching off by default (so no change to the existing code path). @amogkam can you make that change?

**Comment by amogkam** [comment]
@ericl most recent commit sets `batch_ms` to 0 by default

**Review by ericl — APPROVED** [review]

**Pull request merged**